MaxText/maxtext_utils.py

```diff
@@ -92,11 +92,24 @@ def get_train_input_output_trees(func, input_args, input_kwargs):
 # https://arxiv.org/pdf/2204.02311.pdf Appendix B
 def calculate_tflops_training_per_device(num_model_parameters, config, log=True):
   """ Calculate training TFLOP"""
-  learnable_weight_tflops = 6 * num_model_parameters * config.max_target_length * config.per_device_batch_size \
+  if config.use_megatron_tflops:
+    attention_flops = 2 * config.per_device_batch_size * config.max_target_length**2 * config.num_query_heads * config.head_dim
+    ffn1_flops = 2 * config.per_device_batch_size * config.max_target_length * config.mlp_dim * config.emb_dim * len(config.mlp_activations)
+    ffn2_flops = 2 * config.per_device_batch_size * config.max_target_length * config.mlp_dim * config.emb_dim
+    qkv_flops = 6 * config.per_device_batch_size * config.max_target_length * config.emb_dim * config.num_query_heads * config.head_dim
+    proj_flops = 2 * config.per_device_batch_size * config.max_target_length * config.emb_dim * config.num_query_heads * config.head_dim
+
+    # multiplied by 3 to account for fprop and bprop flops
+    learnable_weight_tflops = (ffn1_flops + ffn2_flops + qkv_flops + proj_flops) * config.num_decoder_layers * 3 / 10**12
+    causal_attention_tflops = attention_flops * config.num_decoder_layer * 3 / 10**12
```

Comment:
How does Megatron account for the unprojection matmul? I.e. last layer output to logits.